### PR TITLE
implement #101: margin counter

### PR DESCRIPTION
**Pull request by stefansc1**
count number of times SoC is below desired SoC on departure, added to results report

### CODE REVIEW

File: src/report.py [hunk 1]
```diff
@@ -274,6 +274,11 @@ def aggregate_local_results(scenario, gcID):
         "info": "Number of load cycles per vehicle (averaged)"
     }
 
+    json_results["times below desired soc"] = {
+        "value": scenario.strat.margin_counter,
+        "info": "Number of times vehicle SoC was below desired SoC on departure (without margin)"
```
Comment (jakobgemassmer) [inline src/report.py:279]:
```suggestion
        "info": "Number of times vehicle SoC was below desired SoC on departure"
```

File: src/report.py [hunk 2]
```diff
@@ -274,6 +274,11 @@ def aggregate_local_results(scenario, gcID):
         "info": "Number of load cycles per vehicle (averaged)"
     }
 
+    json_results["times below desired soc"] = {
+        "value": scenario.strat.margin_counter,
```
Comment (jakobgemassmer) [inline src/report.py:278]:
```suggestion
        "value": scenario.strat.desired_counter,
```

File: src/report.py [hunk 3]
```diff
@@ -274,6 +274,11 @@ def aggregate_local_results(scenario, gcID):
...
+        "info": "Number of times vehicle SoC was below desired SoC on departure (without margin)"
+    }
+
```
Comment (jakobgemassmer) [inline src/report.py:281]:
```suggestion
    json_results["times below desired soc with margin"] = {
        "margin": scenario.strat.margin,
        "value": scenario.strat.margin_counter,
        "info": f"Number of times vehicle SoC was below desired SoC on departure "
                f"(with {scenario.strat.margin * 100} % margin)"
    }
```

File: src/strategy.py [hunk 4]
```diff
@@ -52,6 +50,11 @@ def __init__(self, constants, start_time, **kwargs):
         # update optional
         for k, v in kwargs.items():
             setattr(self, k, v)
+        # everything below can not be set by user
+        # for each vehicle, save timestamps when SoC becomes negative
+        self.negative_soc_tracker = {}
+        # count number of times SoC is below desired SoC on departure (used in report)
+        self.margin_counter = 0
```
Comment (jakobgemassmer) [inline src/strategy.py:57]:
```suggestion
        self.desired_counter = 0
        self.margin_counter = 0
```

File: src/strategy.py [hunk 5]
```diff
@@ -123,6 +126,7 @@ def step(self, event_list=[]):
                         # event from the past: simulate optimal charging
                         vehicle.battery.soc = vehicle.desired_soc
                     # check that vehicle has charged enough
+                    self.margin_counter += vehicle.battery.soc < vehicle.desired_soc - self.EPS
```
Comment (jakobgemassmer) [inline src/strategy.py:129]:
```suggestion
                    self.desired_counter += vehicle.battery.soc < vehicle.desired_soc - self.EPS
                    self.margin_counter += vehicle.battery.soc < (1-self.margin)*vehicle.desired_soc - self.EPS

```

File: src/strategy.py [hunk 6]
```diff
@@ -123,6 +126,7 @@ def step(self, event_list=[]):
...
                     if 0 <= vehicle.battery.soc < (1-self.margin)*vehicle.desired_soc - self.EPS:
```
Comment (jakobgemassmer) [inline src/strategy.py:130]: If we add this information to the results json, then we could delete this if statement and its content. Result would be, that the simulation doesn't break if the margin is violated (it still does if SoC becomes negative), but you get the information how often the margin was violated.

**Review by jakobgemassmer — COMMENTED** [review]
General approval, but maybe give the user more information about how "good" the margin is set. And maybe then we don't have to stop the simulation, if the margin is violated.

File: src/report.py (hunk 1 above)
Comment (stefansc1) [reply src/report.py]: separated value into "with margin" and "without margin", updated info according to suggestion below

File: src/report.py (hunk 3 above)
Comment (stefansc1) [reply src/report.py:284]: put both counters into one JSON field

File: src/strategy.py (hunk 6 above)
Comment (stefansc1) [reply src/strategy.py:132]: moved counter into if-block, replaced Exception with Warning. Added final warning at end of simulation if at any time the desired SoC was not reached.

**Review by jakobgemassmer — APPROVED** [review]

**Pull request merged**